### PR TITLE
Hot reloading: Dynamic Channels

### PR DESCRIPTION
**Pull request by JoshVanL**
Branched from #2478 

In order to achieve [hot reloading](#1172), channel callers need to fetch the current channel dynamically. 

PR updates all callers to fetch the current channel type from `channels` during invocation to ensure they are using the latest configured.

/cc @yaron2

### CODE REVIEW

File: pkg/actors/internal/placement_service.go [hunk 1]
```diff
@@ -22,7 +22,7 @@ import (
 type PlacementService interface {
 	io.Closer
 
-	Start(ctx context.Context) error
+	Start(context.Context)
```
Comment (ItalyPaleAle) [inline pkg/actors/internal/placement_service.go:25]: Please leave the error. This was added recently and it helps with future expansion

```suggestion
	Start(ctx context.Context) error
```

File: pkg/actors/placement/client.go [hunk 2]
```diff
@@ -44,43 +44,34 @@ type placementClient struct {
...
 // stream with the connected stream.
-func (c *placementClient) connectToServer(serverAddr string) error {
+func (c *placementClient) connectToServer(ctx context.Context, serverAddr string) error {
 	opts, err := c.getGrpcOpts()
 	if err != nil {
 		return err
 	}
 
-	conn, err := grpc.Dial(serverAddr, opts...)
+	conn, err := grpc.DialContext(ctx, serverAddr, opts...)
```
Comment (ItalyPaleAle) [inline pkg/actors/placement/client.go:57]: If you use DialContext, it's important that you add the `WithBlock` option, since DialContext is non-blocking: https://pkg.go.dev/google.golang.org/grpc#DialContext

File: pkg/actors/placement/placement.go [hunk 3]
```diff
@@ -142,12 +142,12 @@ func (p *actorPlacement) AddHostedActorType(actorType string) error {
 
 // Start connects placement service to register to membership and send heartbeat
 // to report the current member status periodically.
-func (p *actorPlacement) Start(ctx context.Context) error {
+func (p *actorPlacement) Start(ctx context.Context) {
```
Comment (ItalyPaleAle) [inline pkg/actors/placement/placement.go:145]: Let's keep the errors here so the interface can return errors. It was added recently to help with future expansion.


```suggestion
func (p *actorPlacement) Start(ctx context.Context) error {
```

File: pkg/actors/placement/placement.go [hunk 4]
```diff
@@ -142,12 +142,12 @@ func (p *actorPlacement) AddHostedActorType(actorType string) error {
...
 	p.serverIndex.Store(0)
 	p.shutdown.Store(false)
 
-	if !p.establishStreamConn() {
-		return nil
+	if !p.establishStreamConn(ctx) {
+		return
```
Comment (ItalyPaleAle) [inline pkg/actors/placement/placement.go:150]:
```suggestion
		return nil
```

File: pkg/actors/placement/placement.go [hunk 5]
```diff
@@ -241,8 +241,6 @@ func (p *actorPlacement) Start(ctx context.Context) error {
 			}
 		}
 	}()
-
-	return nil
 }
```
Comment (ItalyPaleAle) [inline pkg/actors/placement/placement.go:244]:
```suggestion
	}()

    return nil
}
```

File: pkg/actors/placement/placement_test.go [hunk 6]
```diff
@@ -137,7 +137,7 @@ func TestAppHealthyStatus(t *testing.T) {
 	}).(*actorPlacement)
 
 	// act
-	require.NoError(t, testPlacement.Start(context.Background()))
+	testPlacement.Start(context.Background())
```
Comment (ItalyPaleAle) [inline pkg/actors/placement/placement_test.go:140]:
```suggestion
	require.NoError(t, testPlacement.Start(context.Background()))
```

File: pkg/actors/placement/placement_test.go [hunk 7]
```diff
@@ -84,7 +84,7 @@ func TestPlacementStream_RoundRobin(t *testing.T) {
 		testSrv[leaderServer[0]].setLeader(true)
 
 		// act
-		require.NoError(t, testPlacement.Start(context.Background()))
+		testPlacement.Start(context.Background())
```
Comment (ItalyPaleAle) [inline pkg/actors/placement/placement_test.go:87]:
```suggestion
		require.NoError(t, testPlacement.Start(context.Background()))
```

File: pkg/actors/actors.go [hunk 8]
```diff
@@ -77,8 +77,8 @@ var (
 //nolint:interfacebloat
 type Actors interface {
 	Call(ctx context.Context, req *invokev1.InvokeMethodRequest) (*invokev1.InvokeMethodResponse, error)
-	Init() error
-	Stop()
+	Init(context.Context) error
+	Stop() error
```
Comment (ItalyPaleAle) [inline pkg/actors/actors.go:81]: Sae, please keep errors here

```suggestion
	Stop() error
```

File: pkg/actors/actors.go [hunk 9]
```diff
@@ -239,13 +241,20 @@ func (a *actorsRuntime) Init() error {
 			},
 			AfterTableUpdateFn: func() {
 				a.drainRebalancedActors()
-				a.actorsReminders.OnPlacementTablesUpdated(context.TODO())
+				a.actorsReminders.OnPlacementTablesUpdated(ctx)
```
Comment (ItalyPaleAle) [inline pkg/actors/actors.go:244]: This cannot be `ctx` and must be a background context, because this callback is invoked periodically in background by the placement service (when placement tables are updated)

For now, maybe context.TODO()?


```suggestion
				a.actorsReminders.OnPlacementTablesUpdated(context.TODO())
```

Comment (ItalyPaleAle) [reply pkg/actors/actors.go:244]: Oh, looks like the context is not just for init but for the entire lifecycle of actors? Please confirm if that's the case

File: pkg/actors/actors_test.go [hunk 10]
```diff
@@ -966,7 +966,7 @@ func TestActorsAppHealthCheck(t *testing.T) {
 	clock := testActorsRuntime.clock.(*clocktesting.FakeClock)
 
 	testActorsRuntime.actorsConfig.Config.HostedActorTypes = internal.NewHostedActors([]string{"actor1"})
-	go testActorsRuntime.startAppHealthCheck(
+	go testActorsRuntime.startAppHealthCheck(context.Background(),
```
Comment (ItalyPaleAle) [inline pkg/actors/actors_test.go:969]: Should this context be canceled when tests end?

File: pkg/grpc/server.go [hunk 11]
```diff
@@ -179,14 +187,19 @@ func (s *server) StartNonBlocking() error {
...
 	for _, server := range s.servers {
 		// This calls `Close()` on the underlying listener.
-		server.GracefulStop()
+		go server.GracefulStop()
 	}
 
 	if s.api != nil {
 		if closer, ok := s.api.(io.Closer); ok {
-			closer.Close()
+			return closer.Close()
```
Comment (ItalyPaleAle) [inline pkg/grpc/server.go:202]: I would add an error check like "if err != nil"...

I can see a day when someone adds something to the Close() method after this, and doesn't realize that there's a `return` above.

File: pkg/grpc/server.go [hunk 12]
```diff
@@ -313,22 +330,28 @@ func (s *server) startWorkloadCertRotation() {
 	s.logger.Infof("starting workload cert expiry watcher. current cert expires on: %s", s.signedCert.Expiry.String())
 
 	ticker := time.NewTicker(certWatchInterval)
```
Comment (ItalyPaleAle) [inline pkg/grpc/server.go:332]: Unrelated from this PR but why are we checking if the cert is expired on a regular interval? We know when the cert expires, so we should simply wait until that needs to be renewed, no? That avoids getting locks for no reason.

File: pkg/http/api_directmessaging.go [hunk 13]
```diff
@@ -198,7 +198,7 @@ func (a *api) onDirectMessage(w http.ResponseWriter, r *http.Request) {
 // 1. HTTP header 'dapr-app-id' (path is method)
 // 2. Basic auth header: `http://dapr-app-id:<service-id>@localhost:3500/<method>`
 // 3. URL parameter: `http://localhost:3500/v1.0/invoke/<app-id>/method/<method>`
-func findTargetIDAndMethod(reqPath string, headers http.Header) (targetID string, method string) {
+func findTargetIDAndMethod(reqPath string, headers http.Header) (string, string) {
```
Comment (ItalyPaleAle) [inline pkg/http/api_directmessaging.go:201]: Why was this changed? It seems functionally equivalent in all ways?

File: pkg/runtime/wfengine/wfengine_test.go [hunk 14]
```diff
@@ -64,8 +64,7 @@ func (*mockPlacement) LookupActor(actorType string, actorID string) (name string
 }
 
 // Start implements internal.PlacementService
-func (*mockPlacement) Start(ctx context.Context) error {
-	return nil
+func (*mockPlacement) Start(context.Context) {
```
Comment (ItalyPaleAle) [inline pkg/runtime/wfengine/wfengine_test.go:67]:
```suggestion
func (*mockPlacement) Start(ctx context.Context) error {
	return nil
```

File: pkg/runtime/runtime.go [hunk 15]
```diff
@@ -1597,28 +1641,22 @@ func (a *DaprRuntime) authSecretStoreOrDefault(resource meta.Resource) string {
...
-			return
+		// Return
 		case <-ctx.Done():
-			// Return
-			return
-		default:
-			// nop - continue execution
+			return ctx.Err()
+		// prevents overwhelming the OS with open connections
+		case <-time.After(time.Millisecond * 100):
```
Comment (ItalyPaleAle) [inline pkg/runtime/runtime.go:1659]: The difference is that now the first attempt is delayed 100ms. Maybe ove this `select` at the end of the loop?

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: pkg/actors/actors.go (hunk 9 above)
Comment (JoshVanL) [reply pkg/actors/actors.go:244]: That's right- the context coming into `Init` will be the runner manager context which originates from the base signal context.

File: pkg/grpc/server.go (hunk 11 above)
Comment (JoshVanL) [reply pkg/grpc/server.go:202]: Yes, good shout

File: pkg/http/api_directmessaging.go (hunk 13 above)
Comment (JoshVanL) [reply pkg/http/api_directmessaging.go:201]: We should avoid named return variables at all costs (imo), and _especially_ never do naked returns. The linter was complaining to me about the naked returns.

File: pkg/runtime/runtime.go (hunk 15 above)
Comment (JoshVanL) [reply pkg/runtime/runtime.go:1659]: :+1:

File: pkg/grpc/server.go (hunk 12 above)
Comment (ItalyPaleAle) [reply pkg/grpc/server.go:332]: Yeah it was just an unrelated thing. Good that it's getting ripped anyways

**Comment by JoshVanL** [comment]
@ItalyPaleAle I have address your comments in the [base PR](https://github.com/dapr/dapr/pull/6720)

**Comment by ItalyPaleAle** [comment]
Thank you!

File: pkg/grpc/manager/dial_test.go [hunk 16]
```diff
@@ -1,4 +1,4 @@
-package grpc
+package manager
```
Comment (ItalyPaleAle) [inline pkg/grpc/manager/dial_test.go:1]: Missing copyright header

File: pkg/http/api_directmessaging.go (hunk 13 above)
Comment (ItalyPaleAle) [reply pkg/http/api_directmessaging.go:200]: Ok on not using naked returns (agree). I do like named returns here because they're both strings and using named returns helps documenting the signature of the method.

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: pkg/grpc/manager/dial_test.go (hunk 16 above)
Comment (JoshVanL) [reply pkg/grpc/manager/dial_test.go:1]: :+1:

File: pkg/grpc/api_daprinternal.go [hunk 17]
```diff
@@ -35,7 +35,8 @@ import (
 
 // CallLocal is used for internal dapr to dapr calls. It is invoked by another Dapr instance with a request to the local app.
 func (a *api) CallLocal(ctx context.Context, in *internalv1pb.InternalInvokeRequest) (*internalv1pb.InternalInvokeResponse, error) {
-	if a.appChannel == nil {
+	appChannel := a.channels.AppChannel()
```
Comment (yaron2) [inline pkg/grpc/api_daprinternal.go:38]: Is `a.channels` always guaranteed to be non-empty? Question also applies to other places that make use of it

Comment (JoshVanL) [reply pkg/grpc/api_daprinternal.go:38]: Correct, `channels` will always be non-nil (passed through `Options` or equivalent`), however the underlying channel returned from `AppChannel()` may or may not be nil.

Comment (JoshVanL) [reply pkg/grpc/api_daprinternal.go:38]: We create it as one of the first things we do when creating the `DaprRuntime` https://github.com/dapr/dapr/pull/6762/files#diff-e56f9bb947923c6f8bb4c995bf4b9469a0ed117276223d16b177dcb04399a4bcR183

**Review by yaron2 — APPROVED** [review]
